File: day01/main.go
// packages {{{

package main

import (
	"bufio"
	"log"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

// }}}

func main() {

	// open the file {{{

	file, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// }}}
	// read and parse the file line by line into lists {{{

	var leftList, rightList []int
	var scanner *bufio.Scanner

	scanner = bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		fields := strings.Fields(line)
		if len(fields) >= 2 {
			left, err1 := strconv.Atoi(fields[0])
			right, err2 := strconv.Atoi(fields[1])
			if err1 == nil && err2 == nil {
				leftList = append(leftList, left)
				rightList = append(rightList, right)
			}
		}
	}

	// }}}
	// check for errors {{{

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
	if len(leftList) != len(rightList) {
		log.Fatal("lists are not of the same length")
	}

	log.Println("the total length of the two lists: ", len(leftList), len(rightList))

	// }}}
	// calculate the total distance {{{

	sort.Ints(leftList)
	sort.Ints(rightList)

	dist := 0
	for i := 0; i < len(leftList); i++ {
		dist += int(math.Abs(float64(leftList[i] - rightList[i])))
	}
	log.Println("the total distance between the two lists is: ", dist)

	// }}}
	// calculate the similarity score {{{

	count := make(map[int]int)
	for _, num := range rightList {
		count[num]++
	}

	similarity := 0
	for _, num := range leftList {
		similarity += num * count[num]
	}
	log.Println("the similarity score is: ", similarity)

	// }}}

}

File: day01/main.ts
// import deno functions {{{

import { readLines } from "https://deno.land/std@0.206.0/io/read_lines.ts";
import { ensureFile } from "https://deno.land/std@0.206.0/fs/ensure_file.ts";

// }}}
// read file {{{

const filename = "input.txt";
await ensureFile(filename);
const file = await Deno.open(filename, { read: true });

// }}}
// read and parse the file line by line into lists {{{

const leftList: number[] = [];
const rightList: number[] = [];

try {
  for await (const line of readLines(file)) {
    const fields = line.trim().split(/\s+/);
    if (fields.length >= 2) {
      const left = parseInt(fields[0]);
      const right = parseInt(fields[1]);
      if (!isNaN(left) && !isNaN(right)) {
        leftList.push(left);
        rightList.push(right);
      }
    }
  }
} catch (err) {
  console.error("dunno, something bad happened:", err);
  file.close();
  Deno.exit(1);
} finally {
  file.close();
}

// }}}
// check length of lists {{{

if (leftList.length !== rightList.length) {
  console.error("Lists are not of the same length.");
  Deno.exit(1);
}

console.log(`The total length of the two lists: ${leftList.length}, ${rightList.length}`);

// }}}
// calculate total distance {{{

leftList.sort((a, b) => a - b);
rightList.sort((a, b) => a - b);

let totalDistance = 0;
for (let i = 0; i < leftList.length; i++) {
  totalDistance += Math.abs(leftList[i] - rightList[i]);
}

console.log(`The total distance between the two lists is: ${totalDistance}`);

// }}}
// calculate similarity score {{{

const count: Record<number, number> = {};
for (const num of rightList) {
  count[num] = (count[num] || 0) + 1;
}

let similarityScore = 0;
for (const num of leftList) {
  similarityScore += num * (count[num] || 0);
}

console.log(`The similarity score is: ${similarityScore}`);

// }

